symphony/bdk/core/retry/strategy.py

```diff
@@ -0,0 +1,120 @@
+    return is_client_timeout_error(exception)
+
+
+def is_network_or_minor_error(exception: Exception) -> bool:
+    """Checks if the exception is a network issue or an :py:class:`ApiException` minor error (internal + unauthorized + retry)
+    This is the default function used to check if a given exception should lead to a retry
+
+    :param exception: The exception to be checked
+    :return: True if the exception is an :py:class:`ApiException` with status 500, 401, 429 or is a client timeout error
+             False otherwise
+    """
+    if isinstance(exception, ApiException):
+        if exception.status >= 500 or exception.status == 401 or exception.status == 429:
+            return True
+    return is_client_timeout_error(exception)
+
+
+def is_network_or_minor_error_or_client(exception: Exception) -> bool:
+    """Checks if the exception is a network issue or an :py:class:`ApiException` minor error or a client error
+
+    :param exception: The exception to be checked
+    :return: True if the exception is an :py:class:`ApiException` with status 500, 401, 429, 400
+             or is a client timeout error False otherwise
+    """
+    if isinstance(exception, ApiException):
+        if exception.status >= 500 or exception.status == 401 or exception.status == 429 or exception.status == 400:
+            return True
+    return is_client_timeout_error(exception)
+
+
+def authentication_retry(retry_state: RetryCallState):
+    """Authentication retry strategy
+
+    Retry if the raised exception verifies the can_authentication_be_retried predicate
+    Raise an :py:class`AuthUnauthorizedException` if an :py:class`ApiException` with status 401 is thrown
+    Raise all other exception
+    """
+    if retry_state.outcome.failed:
+        exception = retry_state.outcome.exception()
+        if can_authentication_be_retried(exception):
+            return True
+        if is_unauthorized(exception):
+            unauthorized_message = "Service account is not authorized to authenticate. Check if credentials are valid. "
+            raise AuthUnauthorizedError(unauthorized_message, exception) from exception
+        raise exception
+    return False
+
+
+async def refresh_session_if_unauthorized(retry_state: RetryCallState):
+    """Refreshing AuthSession tokens retry strategy
+
+    Retry if the raised exception verifies the is_network_or_minor_error predicate
+    If this exception is a 401 unauthorized error, we refresh the current AuthSession tokens
+    """
+    if retry_state.outcome.failed:
+        exception = retry_state.outcome.exception()
+        if is_network_or_minor_error(exception):
+            if is_unauthorized(exception):
+                service_auth_session: AuthSession = retry_state.args[0]._auth_session
+                await service_auth_session.refresh()
+            return True
+    return False
+
+
+async def read_datafeed_retry(retry_state: RetryCallState):
```

Comment:
should this be defined in the abstractdatafeedloop class (if possible) (since it knows the internal of AbstractDatafeedLoop)

Reply:
I'd rather group all the strategies in a specific retry module. This would keep tests grouped too.

Reply:
ok (we can always refactor later this is internal APIs)